### PR TITLE
Show spark bars in map tooltip

### PR DESCRIPTION
**Pull request by shaahmad**
Should probably be squashed and merged.

**Before:**
image

**After:**
image

**Before:**
image

**After:**
image

Feedback on React design patterns appreciated.

### CODE REVIEW

**Comment by MarcelGerber** [comment]
image

In the above screenshot, this is definitely not showing the data for March 1 (which is the first day for which data for Germany is available at all). URL would be http://localhost:3099/grapher/coronavirus-cfr?tab=map&year=2020-02-28

**Comment by MarcelGerber** [comment]
Also, there are some charts for which this really doesn't make any sense - maybe it should be an option in the configurator?

One example: http://localhost:3099/grapher/public-events-covid?year=2020-05-08

File: charts/MapTab.tsx [hunk 1]
```diff
@@ -259,30 +344,17 @@ class MapWithLegend extends React.Component<MapWithLegendProps> {
...
                         >
                             <span>
-                                {tooltipDatum
-                                    ? this.context.chart.map.data.formatTooltipValue(
-                                          tooltipDatum.value
-                                      )
-                                    : `No data for ${formatYear(
-                                          inputYear as number
-                                      )}`}
+                                {this.generateTooltip(tooltipDatum, inputYear)}
```
Comment (MarcelGerber) [inline charts/MapTab.tsx:347]: I'm thinking if we should make the tooltip a little (functional) React component itself?

File: site/client/covid/CovidBars.tsx [hunk 2]
```diff
@@ -1,102 +1,20 @@
...
     data: T[]
     x: (d: T) => number
     y: (d: T) => number | undefined
     xDomain: [number, number]
-    onHover: (d: T | undefined, index: number | undefined) => void
     currentX?: number
     highlightedX?: number
     renderValue?: (d: T | undefined) => JSX.Element | undefined
+    onHover?: (d: T | undefined, index: number | undefined) => void
 }
```
Comment (MarcelGerber) [inline site/client/covid/CovidBars.tsx:14]: Hm, this really is `SparkBarsProps` and should probably be replaced.

File: charts/SparkBars.tsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            result.push(indexed[i])
+        }
+        return result
+    }
+
+    applyHoverFunctionIfExists = (
+        d: T | undefined,
+        index: number | undefined
+    ) => () => {
+        this.props.onHover && this.props.onHover(d, index)
```
Comment (MarcelGerber) [inline charts/SparkBars.tsx:83]: You can write this is `this.props.onHover?.(d, index)` just as well

File: site/client/css/components/sparkbars.scss [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            flex: 1;
+            align-self: center;
+            padding-top: 10px;
+            padding-bottom: 10px;
+            padding-left: 6px;
+        }
+    }
+}
+
+.covid-table-mobile {
```
Comment (MarcelGerber) [inline site/client/css/components/sparkbars.scss:79]: More general comment: I don't know where those styles were located before, but it's probably good to remove the old ones.

File: site/client/owid.scss [hunk 5]
```diff
@@ -35,6 +38,7 @@
 @import "css/pages/faq.scss";
 @import "css/pages/covidDataExplorer.scss";
 @import "css/pages/covid.scss";
+@import "css/components/sparkbars.scss";
```
[the remark this answers is not in the document]
Comment (shaahmad) [reply site/client/owid.scss:41]: You're referring to just the `css/pages` vs `css/components`, right?

File: site/client/covid/CovidTimeSeriesValue.tsx [hunk 6]
```diff
@@ -13,13 +13,15 @@ export const CovidTimeSeriesValue = ({
     date,
     latest,
     tooltip,
-    className
+    className,
+    formattedDate = undefined
 }: {
     value: string | undefined
-    date: Date | undefined
+    date?: Date | undefined
```
[the remark this answers is not in the document]
Comment (shaahmad) [reply site/client/covid/CovidTimeSeriesValue.tsx:20]: 👌

File: charts/MapTab.tsx [hunk 7]
```diff
@@ -189,15 +191,97 @@ class MapWithLegend extends React.Component<MapWithLegendProps> {
...
+        return lastVal ? this.sparkBarsDatumXAccessor(lastVal) : undefined
+    }
+
+    generateTooltip(
+        tooltipDatum: ChoroplethDatum | undefined,
+        inputYear: number | undefined
+    ) {
+        return tooltipDatum ? (
+            <div className="map-tooltip">
+                <div className="measure--deaths">
```
[the remark this answers is not in the document]
Comment (shaahmad) [reply charts/MapTab.tsx:248]: oops! good catch

File: charts/SparkBars.tsx (hunk 3 above)
Comment (shaahmad) [reply charts/SparkBars.tsx:83]: 👌

File: site/client/owid.scss (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarcelGerber) [reply site/client/owid.scss:41]: Exactly. This is a really minor thing, but let's keep it consistent and clean :)

[the remark this answers is not in the document]
Comment (shaahmad) [reply site/client/owid.scss]: Fixed!

File: site/client/covid/CovidTimeSeriesValue.tsx [hunk 8]
```diff
@@ -13,13 +13,15 @@ export const CovidTimeSeriesValue = ({
     date,
     latest,
     tooltip,
-    className
+    className,
+    formattedDate = undefined
```
[the remark this answers is not in the document]
Comment (shaahmad) [reply site/client/covid/CovidTimeSeriesValue.tsx]: Fixed!

File: charts/MapTab.tsx [hunk 9]
```diff
@@ -189,15 +191,97 @@ class MapWithLegend extends React.Component<MapWithLegendProps> {
...
+
+    @computed get sparkBarsProps(): SparkBarsProps<SparkBarsDatum> {
+        const sparkBarsData = this.sparkBarsData
+        const [start, end] = this.sparkBarsDomain
+
+        return {
+            data: sparkBarsData,
+            x: this.sparkBarsDatumXAccessor,
+            y: (d: SparkBarsDatum) => d.value,
+            xDomain: [start, end]
```
[the remark this answers is not in the document]
Comment (shaahmad) [reply charts/MapTab.tsx]: Fixed!

File: charts/MapTab.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (shaahmad) [reply charts/MapTab.tsx]: Fixed!

File: charts/SparkBars.tsx (hunk 3 above)
Comment (shaahmad) [reply charts/SparkBars.tsx]: Fixed!

**Comment by MarcelGerber** [comment]
This looks much better already, great work!
I've also pushed a very small fix that uses `takeWhile` from `Util.ts` instead of directly importing lodash.

**Comment by shaahmad** [comment]
> This looks much better already, great work!
> I've also pushed a very small fix that uses `takeWhile` from `Util.ts` instead of directly importing lodash.

Great, thanks!

> In the above screenshot, this is definitely not showing the data for March 1 (which is the first day for which data for Germany is available at all). URL would be http://localhost:3099/grapher/coronavirus-cfr?tab=map&year=2020-02-28

Should be fixed now.

File: charts/MapTab.tsx (hunk 1 above)
Comment (shaahmad) [reply charts/MapTab.tsx]: resolved!

File: site/client/covid/CovidBars.tsx (hunk 2 above)
Comment (shaahmad) [reply site/client/covid/CovidBars.tsx]: resolved!

**Comment by shaahmad** [comment]
> Also, there are some charts for which this really doesn't make any sense - maybe it should be an option in the configurator?
> 
> One example: http://localhost:3099/grapher/public-events-covid?year=2020-05-08

Resolved by checking if country is clickable on the map. I assume unclickable -> we don't want to show in-depth trend -> we shouldn't show sparkbar. I also assume the inverse.

**Comment by MarcelGerber** [comment]
> Resolved by checking if country is clickable on the map. I assume unclickable -> we don't want to show in-depth trend -> we shouldn't show sparkbar. I also assume the inverse. Please let me know if you think this assumption is unsafe.

Yep, I think that's a good way to go.

I also saw your commit removing the duplication between SparkBars and CovidBars, and now that CovidBars is only 11 lines long I think we can drop it altogether. It is only used in `CovidTableColumns.tsx` in 3 places, so you can just replace these three references and drop the old file.


Will take a closer look at your new code soon 🔎, but great job already!

**Comment by shaahmad** [comment]
> CovidBars is only 11 lines long I think we can drop it altogether.

Thanks! Done.

**Comment by MarcelGerber** [comment]
I've found another edge case:
image

(http://localhost:3099/grapher/coronavirus-cfr?tab=map&year=2020-02-22&region=Europe)

The chart is set to Feb 22, but the tooltip shows the data for Feb 24 since that's the earliest data available for Italy. The SparkBars should match the value displayed to the right, so they should also show a bar of 1.51% in this cae.

**Comment by shaahmad** [comment]
> I've found another edge case:
> image
> 
> (http://localhost:3099/grapher/coronavirus-cfr?tab=map&year=2020-02-22&region=Europe)
> 
> The chart is set to Feb 22, but the tooltip shows the data for Feb 24 since that's the earliest data available for Italy. The SparkBars should match the value displayed to the right, so they should also show a bar of 1.51% in this case.

Ah, that is a very weird bug and one that I believe precedes my PR. Really, no data value should be here. It should say "No data for Feb 22". This same bug appears on prod. I will investigate.

**Comment by shaahmad** [comment]
> > The chart is set to Feb 22, but the tooltip shows the data for Feb 24 since that's the earliest data available for Italy. The SparkBars should match the value displayed to the right, so they should also show a bar of 1.51% in this case.
> 
> Ah, that is a very weird bug and one that I believe precedes my PR. Really, no data value should be here. It should say "No data for Feb 22". This same bug appears on prod. I will investigate.

Ah ok, this is related to chart "tolerances" set by authors. I will fix to make the data match the tolerance-adjusted date and push.

**Comment by MarcelGerber** [comment]
I've also noticed that the first data point is always 0, like here:
image

**Comment by shaahmad** [comment]
First data point fixed. I've been staring at these bars for long I didn't even notice it!
Thanks for the CSS adjustments!

**Pull request merged**